### DISCUSSION TITLE
How to add middleware types provided by module?

### DISCUSSION THREAD

**Discussion by Q16solver** [nuxt/nuxt#18462] [Questions] [answered]
I’m currently using a module for auth, sidebase/auth, and it provides an auth middleware for me to use, however it doesn’t seem like providing types for that is supported since nuxt 3 doesn’t detect the middleware that’s been added, is there anyway to add types from the module library?

**Comment by TheAlexLichter** [comment]
I think this rather belongs in the https://github.com/sidebase/nuxt-auth repo ☺️

What types exactly are not available though? The module changes e.g. the `PageMeta` types directly and also [adds the types to nuxt](https://github.com/sidebase/nuxt-auth/blob/main/src/module.ts#L179-L181)

**Reply by Q16solver** [reply]
@manniL Yes.. but the PageMeta types that are autogenerated by nuxt 3 (ref issue: https://github.com/sidebase/nuxt-auth/issues/157) does not seem easy to overwrite with module augmentation, I’m not sure if there’s a simpler way to do this

**Reply by TheAlexLichter** [reply]
Okay, will follow-up there :)

**Reply by Q16solver** [reply]
I believe this is still a wider problem though. There doesn’t seem to be a standard solution in providing types when module libraries define their own middleware for users to use, perhaps we can find a standard solution and document it?

**Reply by Q16solver** [reply]
@manniL There is already some documentation regarding this https://nuxt.com/docs/guide/directory-structure/pages/#typing-custom-metadata though I believe this was actually intended for users and not module lib authors.
However, it is only simple if it’s extra custom metadata with keys that aren’t used by nuxt 3, if we try to override ‘middleware’ key in the type, it’s not going to keep the dynamic type from scanning the files in the middleware directory unless we do some hacky things I think

**Reply by TheAlexLichter** [reply]
> I believe this is still a wider problem though. There doesn’t seem to be a standard solution in providing types when module libraries define their own middleware for users to use, perhaps we can find a standard solution and document it?

Absolutely! Could you raise an issue for that?

**Comment by Q16solver** [comment, marked as answer]
Raised issue https://github.com/nuxt/nuxt/issues/18463, closing discussion as answered
